streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -204,16 +206,17 @@ public void register(final StateStore store,
                                                              final List<ConsumerRecord<byte[], byte[]>> records) {
         final long limit = offsetLimit(storePartition);
         List<ConsumerRecord<byte[], byte[]>> remainingRecords = null;
+        final List<KeyValue<byte[], byte[]>> restoreRecords = new ArrayList<>();
 
         // restore states from changelog records
-        final StateRestoreCallback restoreCallback = restoreCallbacks.get(storePartition.topic());
+        final BatchingStateRestoreCallback restoreCallback = getBatchingRestoreCallback(restoreCallbacks.get(storePartition.topic()));
 
         long lastOffset = -1L;
         int count = 0;
         for (final ConsumerRecord<byte[], byte[]> record : records) {
             if (record.offset() < limit) {
                 try {
-                    restoreCallback.restore(record.key(), record.value());
+                    restoreRecords.add(KeyValue.pair(record.key(), record.value()));
```

Comment:
A `RocksDBException` can be thrown during the restoration process.  The current error handing flow is we catch it in the `RocksDBStateStore`, and wrap it in a `ProcessorStateException` exception and re-throw.  We'll catch it again here, add some contextual information and re-throw for higher levels of the call handle it.  If we want to re-consider how we handle these restoration errors IMHO, it shouldn't block this PR, but add a follow on one to explore our options.